accepted/future-releases/0546-patterns/feature-specification.md

```diff
@@ -621,18 +621,32 @@ It is a compile-time error if:
     expressions in a future release without it being a breaking change, similar
     to default values in parameter lists.*
 
-*   Any two keys in the map are identical. *Duplicate keys are likely to be a
-    copy/paste error. If you want to match the value associated with some key
-    against multiple patterns, you can always use an `||` pattern.*s
-
-*   Any two record keys which both have primitive equality are equal. *Since
-    records don't have defined identity, we can't use the previous rule to
-    detect identical records. But records do support an equality test known at
-    compile time if all of their fields do, so we use that.*
+*   Any two keys in the map are *equivalent constants* (see below). *Duplicate
+    keys are likely to be a copy/paste error. If you want to match the value
+    associated with some key against multiple patterns, you can always use an
+    `||` pattern.*s
 
 *Note that `mapPatternEntries` is not optional, which means it is an error for
 a map pattern to be empty.*
 
+#### Equivalent constants
+
+We can't simply rely on identity to see if two key constants are the same
+because records don't have defined identity. Instead, we specify that two
+values are *equivalent constants* if:
+
+*   They are both not records and are identical.
+
+*   They are both not records, have primitive equality, and are equal according
+    to that primitive equality. *All types with primitive equality have
+    symmetric `==` operators, so it doesn't matter which order the operands of
+    `==` are in.*
```

Comment:
Actually, I was just saying that line 624 could be

    Any two keys in the map are structurally equivalent.

plus a link to the records feature specification where 'structurally equivalent' is defined, and then we wouldn't need this section.

Reply:
ACK, makes sense. Which again means that caring about "primitive equality" here was a red-herring.

Primitive equality is important in places where we are specified as using `==`, but we want to do comparison at compile-time, so we restrict the compile-time comparisons to things where `==` means `identical`.

Here we don't restrict to objects with primitive equality. We allow other objects too, we just only check those for identity. The idea that we could use `==` for objects with primitive equality instead, and get a *better* result, was a mistake, because having primitive equality means we get the *same* result as using `identical`.
So, all in all, just use `identical`.

And well spotted that we already have the definition of "identical-ish + records" that does what we need.

Reply:
Good suggestion, thanks @eernstg!